server/src/main/java/org/apache/druid/segment/indexing/granularity/UniformGranularitySpec.java

```diff
@@ -45,6 +47,15 @@ public UniformGranularitySpec(
     super(inputIntervals, rollup);
     this.queryGranularity = queryGranularity == null ? DEFAULT_QUERY_GRANULARITY : queryGranularity;
     this.segmentGranularity = segmentGranularity == null ? DEFAULT_SEGMENT_GRANULARITY : segmentGranularity;
+    if (Granularity.IS_FINER_THAN.compare(this.segmentGranularity, this.queryGranularity) < 0) {
+      throw DruidException.forPersona(DruidException.Persona.USER)
+                          .ofCategory(DruidException.Category.UNSUPPORTED)
+                          .build(
+                              "SegmentGranularity[%s] must not be finer than QueryGranularity[%s].",
+                              this.segmentGranularity,
+                              this.queryGranularity
+                          );
```

Comment:
`UNSUPPORTED` is mapped to a 5xx error code. We can return an invalid input in this case:
```suggestion
      throw InvalidInput.exception(
         "segmentGranularity[%s] cannot be finer than queryGranularity[%s].", 
          this.segmentGranularity, this.queryGranularity
      );
```
also, it'd be better to capitalize the properties in the error message as a user would specify in the spec.

Reply:
Done.
As for the capitalization, the string representation of Granularities.SECOND is not `SECOND`. It is `{type=period, period=PT1S, timeZone=UTC, origin=null}`.

Reply:
Ah, sorry, I meant capitalization of `segmentGranularity` vs `SegmentGranularity`. Ditto for `queryGranularity`. Basically json annotations for the properties as they would be specified in the spec. Please see related suggestion